corehq/apps/couch_sql_migration/casediff.py

```diff
@@ -611,6 +617,26 @@ def clean_break(self, value):
         self.queue.clean_break = value
 
 
+def prune_premature_diffs(couch_cases, statedb):
+    n_forms = 0
+    case_ids = []
+    cases_by_id = {c.case_id: c for c in couch_cases}
+    stock_forms = get_stock_forms_by_case_id(list(cases_by_id))
+    recs = [CaseRecord(c, stock_forms.get(c.case_id, [])) for c in couch_cases]
+    for case_id, total_forms, processed_forms in statedb.update_cases(recs):
+        if total_forms <= processed_forms:
```

Comment:
Can you describe this to me in terms of what is the "originally calculated" number of forms (possibly calculated weeks ago) and what is the current number of forms? Or maybe this isn't where it's happening; I'm a big confused what's getting filtered out here, and whether that originally calculated number of forms is getting updated

Reply:
A new value of `total_forms` is calculated in the constructor of `CaseRecord` (line 625). The case is _usually freshly loaded from couch_*. `processed_forms` is the number of forms touching this case that have been migrated so far. This is the same condition used enqueue a case to be diffed:

https://github.com/dimagi/commcare-hq/blob/f3a9efe00c916e2e25ccca9d2fc03c7900641d39/corehq/apps/couch_sql_migration/casediff.py#L150-L151

*The exception to _usually freshly loaded from couch_ is when a stale version of the case was present in the LRU cache (not a likely scenario), in which case there is a very slight chance of a false negative (no diff when there should be a diff) if the stale version of the case compares equal to the newer version in SQL, and the newer version in Couch differs. This edge case seems sufficiently unlikely to ignore.